api_generator/src/generator/mod.rs

```diff
@@ -409,94 +422,84 @@ pub fn generate(
             format!("{}.rs", name).as_str(),
+            &mut written_files,
         )?;
     }
 
     // generate functions on root of client
-    let root = code_gen::root::generate(&api, &docs_dir)?;
-    write_file(root, generated_dir, "root.rs")?;
-
-    let generated_modules = fs::read_dir(generated_dir)?
-        .map(|r| {
-            let path = r.unwrap().path();
-            format!("pub mod {};", path.file_stem().unwrap().to_string_lossy())
-        })
+    let mut root = code_gen::root::generate(&api, &docs_dir)?;
+    root.push_str(
+        r#"
+
+mod bulk;
+pub use bulk::*;
+    "#,
+    );
+    write_file(root, None, generated_dir, "root/mod.rs", &mut written_files)?;
+
+    // declare namespace modules in the top-level lib.rs
+    let mods = api
+        .namespaces
+        .keys()
+        .map(|ns| format!("pub mod {};", ns))
         .collect::<Vec<_>>()
         .join("\n");
 
-    write_file(generated_modules, &generated_dir, "mod.rs")?;
-
-    Ok(())
-}
+    let mut sections = HashMap::new();
+    sections.insert("namespace-modules", mods);
+    merge_file(
+        |section| sections.remove(section),
+        generated_dir,
+        "lib.rs",
+        &mut merged_files,
+    )?;
 
-/// Writes the input to the specified file, preceded by a header comment indicating generated code
-fn write_file(input: String, dir: &PathBuf, file: &str) -> Result<(), failure::Error> {
-    let mut generated_path = dir.clone();
-    generated_path.push(file);
-    let path = generated_path.to_string_lossy().into_owned();
+    let mut generated = generated_dir.clone();
+    generated.push(GENERATED_TOML);
 
-    let mut file = File::create(&path)?;
-    file.write_all(
-        "/*
- * Licensed to Elasticsearch B.V. under one or more contributor
- * license agreements. See the NOTICE file distributed with
- * this work for additional information regarding copyright
- * ownership. Elasticsearch B.V. licenses this file to you under
- * the Apache License, Version 2.0 (the \"License\"); you may
- * not use this file except in compliance with the License.
- * You may obtain a copy of the License at
- *
- *	http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing,
- * software distributed under the License is distributed on an
- * \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
- * KIND, either express or implied.  See the License for the
- * specific language governing permissions and limitations
- * under the License.
- */
-// -----------------------------------------------
-// This file is generated, Please do not edit it manually.
-// Run the following in the root of the repo to regenerate:
-//
-// cargo run -p api_generator
-// -----------------------------------------------
-"
-        .as_bytes(),
+    fs::write(
+        generated,
+        toml::to_string_pretty(&GeneratedFiles {
+            written: written_files,
+            merged: merged_files,
+        })?,
```

Comment:
should we make `written_files` and `merged_files` be ordinal ordered, for consistent ordering? I understand they don't need to be, but I ran `cargo make `generate-api` on my machine and the order of entries in `.generated.toml` is different (possible difference in default string ordering implementation on Windows vs. non-Windows)

Reply:
Good point, replaced by a `BTreeSet` in 9ba6ca2 to have a stable ordering.